### PR TITLE
RavenDB-18735: Compression Dictionaries

### PR DESCRIPTION
**Pull request by redknightlois**
cc @maciejaszyk

### CODE REVIEW

**Comment by redknightlois** [comment]
Still triggers without warning:
```
Fatal error. Internal CLR error. (0x80131506)
```

File: bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs [hunk 1]
```diff
@@ -105,6 +105,18 @@ public State(int size)
 
             public Span<byte> EncodingTable => new Span<byte>(_value).Slice(0, _value.Length / 2);
             public Span<byte> DecodingTable => new Span<byte>(_value).Slice(_value.Length / 2);
+
+            public bool CanGrow => false;
+
+            public void Dispose()
+            {
+                throw new NotImplementedException();
```
Comment (ayende) [inline bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs]: ??

File: src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        public Span<byte> DecodingTable => _decodingBuffer.AsSpan(0, _size);
+
+        public bool CanGrow => true;
+
+        public void Grow(int minimumSize)
+        {
+            if (_encodingBuffer.Length < minimumSize)
+            {
+                ArrayPool<byte>.Shared.Return(_encodingBuffer);
+                _encodingBuffer = ArrayPool<byte>.Shared.Rent(minimumSize);
```
Comment (ayende) [inline src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs:35]: Shouldn't you _copy_ the state first?

File: src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            }
+
+            if (_decodingBuffer.Length < minimumSize)
+            {
+                ArrayPool<byte>.Shared.Return(_decodingBuffer);
+                _decodingBuffer = ArrayPool<byte>.Shared.Rent(minimumSize);
+            }
```
Comment (ayende) [inline src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs:42]: Copy the state?

File: src/Sparrow.Server/Compression/Encoder3Gram.cs [hunk 4]
```diff
@@ -363,7 +368,12 @@ private void BuildDictionary(in FastList<SymbolCode> symbolCodeList)
             // We haven't stored it yet. So we are calculating it. 
             int numberOfEntriesInTable = (_state.EncodingTable.Length - 4) / Unsafe.SizeOf<Interval3Gram>();
             if (numberOfEntriesInTable < dictSize)
-                throw new ArgumentException("Not enough memory to store the dictionary");
+            {
+                if (_state.CanGrow)
+                    _state.Grow(dictSize);
+                else
+                    throw new ArgumentException("Not enough memory to store the dictionary");
```
Comment (ayende) [inline src/Sparrow.Server/Compression/Encoder3Gram.cs]: Add how much memory is required? And what is the current size?

File: src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs (hunk 2 above)
Comment (redknightlois) [reply src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs:35]: There is no state when that happens... we know before the fact how big it has to be. Will add some extra documentation to that.

File: src/Sparrow.Server/Compression/Encoder3Gram.cs (hunk 4 above)
Comment (redknightlois) [reply src/Sparrow.Server/Compression/Encoder3Gram.cs]: It is not actually memory is `reserved memory`... we want to limit the work and therefore the size of the actual dictionary. We may relax this over time, for now I want to have control of that.

**Comment by redknightlois** [comment]
There is no reason why not to merge the partial fixes for 2 of the underlying issues surrounding RavenDB-18735. @arekpalinski 
 Feel free to review, merge and close.

File: bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs [hunk 5]
```diff
@@ -105,6 +105,18 @@ public State(int size)
...
+            }
+
+            public void Grow(int minimumSize)
+            {
+                throw new NotImplementedException();
```
Comment (arekpalinski) [inline bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs]: Shouldn't we rather throw NotSupportedException? Same like you do in NativeMemoryEncoderState.cs

File: src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Sparrow.Server/Compression/AdaptiveMemoryEncoderState.cs:35]: Did you add the "extra documentation" you mentioned here?

File: src/Sparrow.Server/Compression/Encoder3Gram.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Sparrow.Server/Compression/Encoder3Gram.cs]: 1. I believe Oren meant to adding some extra info to the exception message so we can have more details when it happens again.
2. `ArgumentException` doesn't seem to be suitable here.

File: bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs (hunk 1 above)
Comment (arekpalinski) [reply bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs]: 1. Do we throw from `Dispose`?
2. Given that we don't see it throwing it means that the `Dispose` isn't called. My concern is that we might not dispose other objects implementing `IEncoderState` which do have something to dispose. Please verify that

File: src/Voron/Data/Containers/Container.cs [hunk 6]
```diff
@@ -402,33 +434,35 @@ public static void Delete(LowLevelTransaction llt, long containerId, long id)
...
             int totalSize = 0, count = 0;
             for (var i = 0; i < entriesOffsets.Length; i++)
             {
                 if(entriesOffsets[i].Size ==0)
                     continue;
                 count++;
                 totalSize += entriesOffsets[i].Size;
-            }
-
+            }
```
Comment (arekpalinski) [inline src/Voron/Data/Containers/Container.cs]: Minor - formatting: Why some spaces were added to the end of lines here and above?

File: test/FastTests/Sparrow/Encoder3GramTests.cs [hunk 7]
```diff
@@ -25,6 +25,15 @@ public State(int size)
...
             public Span<byte> EncodingTable => new Span<byte>(_value).Slice(0, _value.Length / 2);
             public Span<byte> DecodingTable => new Span<byte>(_value).Slice(_value.Length / 2);
+            
+            public bool CanGrow => false;
+
+            public void Dispose(){}
+
+            public void Grow(int minimumSize)
+            {
+                throw new NotSupportedException();
```
Comment (arekpalinski) [inline test/FastTests/Sparrow/Encoder3GramTests.cs]: Please add a message to the exception

**Comment by arekpalinski** [comment]
@redknightlois 

> There is no reason why not to merge the partial fixes for 2 of the underlying issues surrounding RavenDB-18735. @arekpalinski Feel free to review, merge and close.

Please check my comments. Once they are applied I will merge the PR

File: bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs (hunk 5 above)
Comment (redknightlois) [reply bench/Micro.Benchmark/Benchmarks/HopeEncoder.cs]: It's never gonna be tested, it is there for triggering a failure at the user but it is a code sanity thing. Yes.

**Comment by redknightlois** [comment]
> Did you add the "extra documentation" you mentioned here?

Yes it was added, moved it up a bit so it is more prominent.

> Given that we don't see it throwing it means that the Dispose isn't called. My concern is that we might not dispose other objects implementing IEncoderState which do have something to dispose. Please verify that

Yes it is been called, we haven't run the dispose in the benchmark and it would definitely fail if done.

> `ArgumentException` doesn't seem to be suitable here.

It is a borderline case, because `ArgumentException` relates to the table type not supporting growing the backing store. Using `InsuficientMemoryException` now instead.

**Pull request merged**